base/sort.jl

```diff
@@ -1091,77 +1093,109 @@ julia> v[p]
  1
  2
  3
+
+julia> A = [8 7; 5 6]
+2×2 Matrix{Int64}:
+ 8  7
+ 5  6
+
+julia> sortperm(A, dims = 1)
+2×2 Matrix{Int64}:
+ 2  4
+ 1  3
+
+julia> sortperm(A, dims = 2)
+2×2 Matrix{Int64}:
+ 3  1
+ 2  4
 ```
 """
-function sortperm(v::AbstractVector;
-                  alg::Algorithm=DEFAULT_UNSTABLE,
-                  lt=isless,
-                  by=identity,
-                  rev::Union{Bool,Nothing}=nothing,
-                  order::Ordering=Forward)
-    ordr = ord(lt,by,rev,order)
-    if ordr === Forward && isa(v,Vector) && eltype(v)<:Integer
-        n = length(v)
+function sortperm(A::AbstractArray;
+    alg::Algorithm=DEFAULT_UNSTABLE,
+    lt=isless,
+    by=identity,
+    rev::Union{Bool,Nothing}=nothing,
+    order::Ordering=Forward,
+    dims... #to optionally specify dims argument
+)
+    ordr =  ord(lt, by, rev, order)
+    if ordr === Forward && isa(A, Vector) && eltype(A) <: Integer
+        n = length(A)
         if n > 1
-            min, max = extrema(v)
+            min, max = extrema(A)
             (diff, o1) = sub_with_overflow(max, min)
             (rangelen, o2) = add_with_overflow(diff, oneunit(diff))
-            if !o1 && !o2 && rangelen < div(n,2)
-                return sortperm_int_range(v, rangelen, min)
+            if !o1 && !o2 && rangelen < div(n, 2)
+                return sortperm_int_range(A, rangelen, min)
             end
         end
     end
-    ax = axes(v, 1)
-    p = similar(Vector{eltype(ax)}, ax)
-    for (i,ind) in zip(eachindex(p), ax)
-        p[i] = ind
-    end
-    sort!(p, alg, Perm(ordr,v))
+    perm = Perm(
+        ordr,
+        vec(A)
+    )
+    ix = Base.copymutable(LinearIndices(A))
+    sort!(ix; dims..., alg = alg, order=perm)
 end
 
-
 """
-    sortperm!(ix, v; alg::Algorithm=DEFAULT_UNSTABLE, lt=isless, by=identity, rev::Bool=false, order::Ordering=Forward, initialized::Bool=false)
+    sortperm!(ix, A; alg::Algorithm=DEFAULT_UNSTABLE, lt=isless, by=identity, rev::Bool=false, order::Ordering=Forward, initialized::Bool=false, [dims::Integer])
 
-Like [`sortperm`](@ref), but accepts a preallocated index vector `ix`.  If `initialized` is `false`
-(the default), `ix` is initialized to contain the values `1:length(v)`.
+Like [`sortperm`](@ref), but accepts a preallocated index Vector or Array `ix` of the same length and indexing as `ix`.  If `initialized` is `false`
```

Comment:
```suggestion
Like [`sortperm`](@ref), but accepts a preallocated index vector or array `ix` with the same `axes` as `A`.  If `initialized` is `false`
```
Capitalization makes me think of the types `Vector` and `Array`, when we really mean `AbstractVecctor` and `AbstractArray`. We could try to be substantially more clear, or we could just leave it lower case.

"length and indexing" isn't wrong, but we check that the `axes` are equal, so that's probably what we should say we do. `zeros(2,3)` and `zeros(3,2)` have the same length and both use one-based indexing.

`ix` should be `A`